rfcs/0000-nixpkgs-ifd.md

```diff
@@ -0,0 +1,182 @@
+
+3. > IFD, is too controversial, don't do it!
+
+   I think this is a classic example of don't let the perfect be the enemy of the good.
+   The problems with IFD and the problems IFD is trying to address both don't let a lot of attention.
+   The fact of the matter is Nixpkgs is how this community coordinates with itself, and agrees on priorities.
+   If it isn't being used in Nixpkgs, there is hard ceiling of how much attention it will get.
+
+   The benefits of IFD don't get enough attention.
+   A package in Nixpkgs is more than a derivation: there's the `meta` as I mentioned above.
+   There's also being able to read the code and (somewhat) understand what's going in reference to other derivations.
+   Finally, there's being able to `override`, `overrideAttrs`, etc. the derivation downstream.
+   IFD alone allows computed packages that follow all these norms.
+
+4. Say I want IFD that depends on other IFD?
+
+   In other words, can one import a derivation that is itself evaluated from and import from a derivation?
+   No, not without introducing another round of building and evaluating for Hydra.
+   But I don't think we need arbitrarily-deep dynamism anyways: it is a tool that should be used with care anyways, because stasis \[staticism?\] is the goodly disciplinarian that makes Nixpkgs so great.
+
+   That said, `cabal2nix` is written in Haskell, `crate2nix` is written in Rust, etc. etc.
+   We can vendor enough code to build these tools and thus bootstrap the IFD we will do.
+   Per the 3rd rule for Nixpkgs above, as long as we make the vendoring automatic and pure, this is fine, and improvement upon today.
+   Also, even if we didn't have the "one round of dynamism" restriction, we would still have the bootstrapping issue.
+
+# Alternatives
+[alternatives]: #alternatives
+
+1. Instead of evaluating Nixpkgs twice, just evaluate `allImportedDerivations` the first round.
+
+   We could do this, and would reduce total eval time, yes.
+   But, I think it would come at the cost of inciting great controversy.
+   This means users of the `enableImportFromDerivation = false` subset of Nixpkgs would still have to *wait*, for all the IFD to complete first.
+   And remember, with mass rebuilds, that could be quite some time.
+   Increasing the critical path length of *everything* we do with Nixpkgs would cause real pain in some quarters, and I don't want that to pay that as the cost of IFD.
+
+   With the plan as written, users of packages depending on IFD do have to wait slightly longer as the first eval is longer (and we wait for it before beginning to build `allImportedDerivations`).
+   But I think that is fair; we would be the "new constituency", the bottom of the pecking order, and so we should be patient so that other's workflows are not disturbed.
+
+   Longer term we could revisit this, or we could e.g. double down on automatic vendoring, committing all generated code to a second "roll-up" repo.
```

Comment:
One benefit of `allImportedDerivations` could really be that updating the generated code becomes easier because it is equivalent to updating some hard coded inputs (like the hackage data tarball) and then running a `nix-build -A allImportedDerivations.…`.

For better or worse this would also put more scrutiny on the maintenance of `lang2nix` tools. E. g. we'd need to do something about `emacs2nix` which currently is built against a reasonably ancient version of `hnix` we no longer have in nixpkgs.

Reply:
E. g. <https://github.com/NixOS/nixpkgs/issues/150941>